Status of Sender (#73)
AMWA-TV/nmos-sink-metadata-processing: docs/Behaviour - Server Side.md

### Request Methods

- `GET` request returns the last successfully applied Active Constraints.
- `PUT` request MUST be validated and if the Sender supports Parameter Constraints used in the proposed Active Constraints and is capable to adhere at least one of the Constraint Sets, then the associated Inputs may be reconfigured, which may eventually cause the update of the corresponding Flows, Sources, `/transportfile` and Sender. If successful, the operation MUST return the accepted Active Constraints, otherwise an error MUST be returned and nothing changed.
- `DELETE` request MUST clear Active Constraints. It turns the Sender to the initial state which excludes IS-11 out of the process.

## State of Sender

A Sender managed with IS-11 has five states:
- `Unconstrained` when Active Constraints of this Sender is an empty `constraint_sets` array.
- `Constrained` when this Sender, the Flow and the `/transportfile` (if used) associated with this Sender and the Source associated with this Flow satisfy the Active Constraints.

> - `Active Constraints Violation` when this Sender, the Flow or the `/transportfile` (if used) associated with this Sender or the Source associated with this Flow does not satisfy the Active Constraints. When the Sender is in this state, it MUST NOT allow [IS-05][IS-05] activations.

Reviewer: Must not be active ... as It could be active and not allow further activation.

Author: ```suggestion
- `Active Constraints Violation` when this Sender, the Flow or the `/transportfile` (if used) associated with this Sender or the Source associated with this Flow does not satisfy the Active Constraints. When an inactive Sender is in this state, it MUST NOT allow [IS-05][IS-05] activations.
```
Breaking connections of active Senders is described in "Preventing breaking restrictions" (maybe I have to rename it to "Preventing restrictions violation".

- `DELETE` request MUST clear Active Constraints. It turns the Sender to the initial state which excludes IS-11 out of the process.

## Receivers
## State of Sender

A Sender managed with IS-11 has five states:
- `Unconstrained` when Active Constraints of this Sender is an empty `constraint_sets` array.
- `Constrained` when this Sender, the Flow and the `/transportfile` (if used) associated with this Sender and the Source associated with this Flow satisfy the Active Constraints.
- `Active Constraints Violation` when this Sender, the Flow or the `/transportfile` (if used) associated with this Sender or the Source associated with this Flow does not satisfy the Active Constraints. When the Sender is in this state, it MUST NOT allow [IS-05][IS-05] activations.
- `No Signal` when there is no signal from Inputs associated with this Sender.
- `Awaiting Signal` when Active Constraints was just PUT and the signal from Inputs associated with this Sender is not stable yet. This is a transitional state until one of the previous ones can be established.

> If an active Receiver becomes unable to handle Flow, it is REQUIRED to take the following actions:

Reviewer: Should we keep dynamic receiver capabilities violations: Ex. after a receiver started receiving a stream and displaying it on a monitor ... I unplug the 4K monitor and plug an HD monitor ... this will change the capabilities of the receiver which will make the incoming 4K signal in violation of those capabilties .... I guess that the receiver have to stop streaming either because its capabilities changed but mostly because of the violation.

> ## Preventing breaking restrictions

Author: ```suggestion
## Preventing restrictions violation
```

> If Receiver activation is requested with a `/transportfile` breaking Receiver Capabilities, the Receiver is REQUIRED to refuse such request.

Reviewer: and fail PATCH with error 400 ... (awaiting consensus from NMOS Slack channel)

Author: ```suggestion
If `/staged` resource modification is requested for a Receiver via PATCH with a `/transportfile` violating Receiver Capabilities, the Receiver is REQUIRED to refuse such request with `400 Bad Request`. If the Receiver is active, it also MUST stop receiving the Flow and REQUIRED to take the following actions:
- MUST set the [IS-05][IS-05] `/active`'s `master_enable` property to `false`.
- MUST update the [IS-04][IS-04] `subscription`'s `active` property to `false` (regardless of unicast or multicast).
```

Reviewer: need to cover /bulk/receivers endpoint also which uses a POST and a number of receivers  ...

Author: ```suggestion
If staged parameters update is requested for a Receiver with a `/transportfile` violating Receiver Capabilities, the Receiver is REQUIRED to refuse such request with `400 Bad Request`. If the Receiver is active, it also MUST stop receiving the Flow by taking the following actions:
- MUST set the [IS-05][IS-05] `/active`'s `master_enable` property to `false`.
- MUST update the [IS-04][IS-04] `subscription`'s `active` property to `false` (regardless of unicast or multicast).
```

> 

Author: ```suggestion

If an active Receiver changes its Receiver Capabilities, then it MUST validate `/transportfile` against them and if the validation fails, it MUST stop receiving the Flow by taking the following actions:
- MUST set the [IS-05][IS-05] `/active`'s `master_enable` property to `false`.
- MUST update the [IS-04][IS-04] `subscription`'s `active` property to `false` (regardless of unicast or multicast).
```
